hs/src/Reach/Optimize.hs

```diff
@@ -920,11 +926,72 @@ preFloat = \case
     void $ updateVarSet dv $ Just . maybe 1 succ
   _ -> return ()
 
--- Optimize is generally top-down, but this is locally bottom-up. We float after optimizing the tail
+updateFusion :: DLVar -> SwitchCases DLTail -> App ()
+updateFusion dv scs = do
+  fusionCasesR <- asks eFusionCases
+  liftIO $ modifyIORef' fusionCasesR $ M.insert dv scs
+
+readFusionCases :: DLVar -> App (Maybe (SwitchCases DLTail))
+readFusionCases dv = do
+  fusionCases <- liftIO . readIORef =<< asks eFusionCases
+  return $ M.lookup dv fusionCases
+
+preFuse :: DLStmt -> App (M.Map DLVar (SwitchCases DLTail))
+preFuse s = do
+  oldEnv <- liftIO . readIORef =<< asks eFusionCases
+  case s of
+    DL_LocalSwitch _ dv _ -> updateFusion dv mempty
+    DL_Let {} -> cantFuse
+    DL_Var {} -> cantFuse
```

Comment:
You should make a comment that we could if they weren't bound, but that we're not trying to track that yet